### ISSUE TITLE
SQLite driver

### ISSUE THREAD

**Issue by dkotik** [ThreeDotsLabs/watermill#433]
@roblaszczak, @m110 I hope to contribute an SQLite driver some time this summer. I am opening this issue to collect ideas and to hear suggestions from you and the community. Two questions to kick off:

1. I noticed that all the drivers are in separate repositories, and I find it a bit odd as opposed to keeping them together in the same repo with their own `go.mod` file. Another project wouldn't be able to use them as they are too specific. Nonetheless, how would like me to go about implementation? Follow the same pattern?
2. I enjoy working with <https://github.com/zombiezen/go-sqlite> driver. It does not use the standard library sql interfaces and carries SQLite-specific benefits because of it. Also no CGO. Is it ok if I use it for watermill implementation at least initially? I can do an additional standard lib-compatible one later - it should not be hard to re-implement once it is done.

**Comment by roblaszczak** [comment]
> @roblaszczak, @m110 I hope to contribute an SQLite driver some time this summer. I am opening this issue to collect ideas and to hear suggestions from you and the community. Two questions to kick off:
> 
> 1. I noticed that all the drivers are in separate repositories, and I find it a bit odd as opposed to keeping them together in the same repo with their own `go.mod` file. Another project wouldn't be able to use them as they are too specific. Nonetheless, how would like me to go about implementation? Follow the same pattern?

I remember discussing with someone who is all "drivers" in one repo in a similar project, and it's a pain 😅 
It kinda works now when all drivers are separate, so we'll prefer not to make a precedence.

It's a good question if it should be in https://github.com/ThreeDotsLabs/watermill-sql repo or a separate one - we would need to consider tradeoffs.

> 2. I enjoy working with https://github.com/zombiezen/go-sqlite driver. It does not use the standard library sql interfaces and carries SQLite-specific benefits because of it. Also no CGO. Is it ok if I use it for watermill implementation at least initially? I can do an additional standard lib-compatible one later - it should not be hard to re-implement once it is done.


@m110 pointed me to https://github.com/walterwanderley/watermill-sqlite - you may take a look at it as a reference, but as far as I understand, it depends on a driver that depends on CGO?

So https://github.com/zombiezen/go-sqlite sounds like a nice alternative! It would be just perfect to not expose details of that library anywhere - so we could eventually swap the driver in the future if possible.

I'm not sure if you know this link, but it could be useful for you: https://watermill.io/docs/pub-sub-implementing/

**Comment by AlexCuse** [comment]
Ah I see zombiezen package uses the modernc driver.  FWIW I think there's a strong case to use modernc.org driver directly within the primary watermill-sql package for ease of maintenance, but I would be very interested to compare the two using the watermill benchmarks!

**Comment by dkotik** [comment]
I am moving back to Ukraine, will follow up in a few months.

**Comment by dkotik** [comment]
I am back. I looked over Walter's attempt briefly. I do not think he intends to finish it soon. The SQL shows signs that he did not get into particulars of SQLite yet: it does not have TIMESTAMP type. Also, he put the payload as JSON.

I will start pushing commits on Monday starting with the C-less modernc driver. I will add zombiezen as a parallel implementation as its own module.

**Comment by dkotik** [comment]
@AlexCuse, alright: I got the ugly modernc driver version passing my basic tests. It seems to pass most of the standard library PubSub acceptance tests as well. Here is the link: https://github.com/dkotik/watermillsqlite

**Comment by dkotik** [comment]
Progess report: I am passing 11 of the official acceptance tests:

- [x] tests.TestPublishSubscribe
- [x] tests.TestConcurrentSubscribe
- [x] tests.TestConcurrentSubscribeMultipleTopics
- [x] tests.TestResendOnError
- [x] tests.TestContinueAfterSubscribeClose
- [x] tests.TestConcurrentClose
- [x] tests.TestContinueAfterErrors
- [x] tests.TestPublishSubscribeInOrder
- [x] tests.TestPublisherClose
- [x] tests.TestTopic
- [x] tests.TestNewSubscriberReceivesOldMessages
- [ ] tests.TestNoAck
- [ ] tests.TestMessageCtx
- [ ] tests.TestSubscribeCtx // hangs
- [ ] consumer group tests

**Comment by AlexCuse** [comment]
This is great I can see a lot of uses for the sqlite backend on remote-deployed machines.

**Comment by dkotik** [comment]
Still rough and flaky, but I tagged v0.0.1 on

```sh
go get -u github.com/dkotik/watermillsqlite/wmsqlitemodernc
```

**Comment by AlexCuse** [comment]
I wonder if there is much value in keeping the stdlib implementation (maybe could be folded into [watermill-sql ](https://github.com/threedotslabs/watermill-sql) as another supported DB somehow?)

If zombiezen package is faster and more stable making that the default or only option for a dedicated watermill-sqlite driver makes a lot of sense to me.
